launcher/modplatform/modrinth/ModrinthPackExportTask.cpp

```diff
@@ -0,0 +1,309 @@
+const QStringList ModrinthPackExportTask::FILE_EXTENSIONS({ "jar", "litemod", "zip" });
+
+ModrinthPackExportTask::ModrinthPackExportTask(const QString& name,
+                                               const QString& version,
+                                               const QString& summary,
+                                               InstancePtr instance,
+                                               const QString& output,
+                                               MMCZip::FilterFunction filter)
+    : name(name)
+    , version(version)
+    , summary(summary)
+    , instance(instance)
+    , mcInstance(dynamic_cast<const MinecraftInstance*>(instance.get()))
+    , gameRoot(instance->gameRoot())
+    , output(output)
+    , filter(filter)
+{}
+
+void ModrinthPackExportTask::executeTask()
+{
+    setStatus(tr("Searching for files..."));
+    setProgress(0, 0);
+    collectFiles();
+}
+
+bool ModrinthPackExportTask::abort()
+{
+    if (task != nullptr) {
+        task->abort();
+        task = nullptr;
+        emitAborted();
+        return true;
+    }
+
+    pendingAbort = true;
+    return true;
+}
+
+void ModrinthPackExportTask::collectFiles()
+{
+    files.clear();
+    if (!MMCZip::collectFileListRecursively(instance->gameRoot(), nullptr, &files, filter)) {
+        emitFailed(tr("Could not search for files"));
+        return;
+    }
+
+    pendingHashes.clear();
+    resolvedFiles.clear();
+
+    if (mcInstance) {
+        mcInstance->loaderModList()->update();
+        connect(mcInstance->loaderModList().get(), &ModFolderModel::updateFinished, this, &ModrinthPackExportTask::collectHashes);
+    } else
+        collectHashes();
+}
+
+void ModrinthPackExportTask::collectHashes()
+{
+    for (const QFileInfo& file : files) {
+        const QString relative = gameRoot.relativeFilePath(file.absoluteFilePath());
+        // require sensible file types
+        if (!std::any_of(PREFIXES.begin(), PREFIXES.end(),
+                         [&relative](const QString& prefix) { return relative.startsWith(prefix + QDir::separator()); }))
+            continue;
+        if (!std::any_of(FILE_EXTENSIONS.begin(), FILE_EXTENSIONS.end(), [&relative](const QString& extension) {
+                return relative.endsWith('.' + extension) || relative.endsWith('.' + extension + ".disabled");
+            })) {
+            continue;
+        }
+
+        QCryptographicHash sha512(QCryptographicHash::Algorithm::Sha512);
+
+        QFile openFile(file.absoluteFilePath());
+        if (!openFile.open(QFile::ReadOnly)) {
+            qWarning() << "Could not open" << file << "for hashing";
+            continue;
+        }
+
+        const QByteArray data = openFile.readAll();
+        if (openFile.error() != QFileDevice::NoError) {
+            qWarning() << "Could not read" << file;
+            continue;
+        }
+        sha512.addData(data);
```

Comment:
can this use [`Hasher::ModrinthHasher`](https://github.com/PrismLauncher/PrismLauncher/blob/ed085630db63401ac50d26403d953f5d9993b9ce/launcher/modplatform/helpers/HashUtils.h#L36) instead? maybe that can also be changed to progressively get the Sha512 hash and then the Sha1 hashes on subsequent `executeTask` calls or something, so that you can also substitute the other hash computation down there, it would be really sweet i think :3

Reply:
i'm not sure about this tbh... changing the behaviour based on number of calls doesn't sound the best

Reply:
Oh, OK... i just don't know if this would break existing code

Reply:
Having looked at it more (sorry for the comment spam), it looks like it would be innefficient since it's reopening the file multiple times, and the different platform hashers would be inconsistent with each other. I don't think it's practical in this pr, sorry (quite a lot of refactoring would probably be required) :'(

Reply:
> Oh, OK... i just don't know if this would break existing code

It really shouldn't. As I said, behavior when re-executing tasks shouldn't be expected to work as intended by default. If something were to break from that change, the code that broke is wrong, not your change.

> it looks like it would be innefficient since it's reopening the file multiple times

I don't see that mattering a lot tbh, most jars are relatively small, the OS would most likely cache the reads, and even if worst came to worst, it would be really trivial to cache the read in `Hasher`.

well, i'm not gonna force you to make this change if you really don't want to, but I do think having this exact hash calculation type of thing all over the place in the code base is not a good architecture to follow, which is why I made `Hasher` in the first place. It's not the end of the world however, don't worry about it. thanks for the patience dealing with my reviewer annoyance :-)

Reply:
you weren't annoying :)
And my decision was mainly because of the other reasons